tests/cpp/test_multidevice_overlap.cpp

```diff
@@ -0,0 +1,283 @@
+//        - B[K,N] sharded row-wise:
+//          locally of size [1, K/num_devices, N]
+//     output:
+//        - C[M,N]=matmul(A,B), sharded on dimension M:
+//          dimension M is split by the factor `num_devices`
+//          so C is viewed as [num_devices, M/num_devices,N]
+//          and the allocation size of M is [1, M/num_devices,N]
+// Up to some broadcast and view ops, a straightforward program to generate the
+// output could be summarized as
+//     | C_unreduced = pointwise_multiply(A,B)
+//     | C_locally_reduce = reduction(C_unreduced, axis=`K/num_devices`, op=sum)
+//     | C = reduce_scatter(C_unreduced, op=sum)
+// where:
+// - C has unsharded size [M,num_devices,K/num_devices,N],
+//    and is sharded on `num_devices`
+// - C_locally_reduce has unsharded size [M,num_devices,N],
+//    and is sharded on `num_devices`
+// - C has unsharded size [num_devices, M/num_devices, N]
+//    and is sharded on `num_devices`
+//
+// We want to compare this baseline program with one that is functionnally
+// identical but achieves more overlap between computations and communications.
+// Our goal is to interlave the comms and compute using a technic called
+// "reduce-scatter based pipelining". To do so, we further split the row
+// dimension M with a factor `S` representing the number of tiles, and we apply
+// the operations successively on tensors slices accross S, changing stream at
+// each iteration. Assuming the following shapes:
+//     - A [S, num_devices, M/S, K/num_devices], sharded on num_devices
+//     - B [num_devices, K/num_devices, N], sharded on num_devices
+//     - C [S, num_devices, M/(S*num_devices), N], sharded on num_devices
+// the program could be summarized as:
+//     | for (j=0; j<S; j++) {
+//     |   setCurrentStream(Stream[j])
+//     |   C_unreduced[j] = pointwise_multiply(A[j],B)
+//     |   C_locally_reduce[j] = local_reduction(C_unreduced[j], axis=`K/num_devices`, op=sum)
+//     |   C[j]=reduce_scatter(C_locally_reduce[j], op=sum)
+//     | }
+// where "[j]" referes to taking a slice onto the `S` dimension.
+// Remarks:
+//   1) it is convenient to have "S" as being the outermost dimension so
+//      C_locally_reduce[j] is a contiguous buffer.
+//   2) The layout needs to match
+//      the reduce-scatter semantics, i.e., the first dimension is reduced and
+//      the second is scattered. This is why we choose the layouts to be
+//      [S, sharded_axis, M, ...]
+// clang-format on
+TEST_F(OverlapTest, SimpleComputeComm) {
+  std::vector<c10::cuda::CUDAStream> streams;
+  for (auto j : c10::irange(params.S)) {
+    // define the sliced tensors
+    auto ta_j = getSlice(ta, j);
+    auto tc_unreduced_j = getSlice(tc_unreduced, j);
+    auto tc_locally_reduced_j = getSlice(tc_locally_reduced, j);
+    auto tc_j = getSlice(tc, j);
+
+    if (params.use_different_streams) {
+      auto new_stream = c10::cuda::getStreamFromPool(
+          /*isHighPriority=*/true, my_device_index);
```

Comment:
Can you comment on the high-priority part?

Reply:
To give more details: 
- CUDA provides a notion of stream priority: https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1ge2be9e9858849bf62ba4a8b66d1c3540
- Torch have two pools of streams, one with "high priority", the other with "low priority". https://pytorch.org/cppdocs/notes/tensor_cuda_stream.html
- in Fuser, for now, it is not clear how we could leverage the stream priorities, so we decide in the meantime to only use high priority streams.

Reply:
Nit: I tend to use default unless the non-default value is proven to be important. In this case, the benefit of using a high-priority stream seems unclear.

Reply:
alright, that makes sense. Let me fix it

Reply:
fixed in https://github.com/NVIDIA/Fuser/pull/2475